#!/bin/bash
#$ -V
#$ -cwd
#$ -S /bin/bash
#$ -N mitobim
#$ -o $JOB_NAME.o$JOB_ID
#$ -e $JOB_NAME.e$JOB_ID
#$ -q raycc,ray512cc
#$ -pe fill 1
#$ -P communitycluster

#This file will process raw Illumina data using Trimmomatic.  This will be followed by mapping to a reference mitochondrial genome using MIRA4 to create a new mitochondrial genome assembly.

BASEDIR=/lustre/scratch/daray/mitobim
WORKDIR=$BASEDIR/output

mkdir $WORKDIR
cd $WORKDIR

######
#set up variables - will vary according to your needs, files and file locations
######
ABBREV=?????  #This will be the name you use to process your files.  You will need to change the RP1, RP2, RU1, and RU2 slots just below here accordingly.
#These files will be generated by trimmomatic
RP1=$ABBREV"_R1_paired.fastq"
RP2=$ABBREV"_R2_paired.fastq"
RU1=$ABBREV"_R1_unpaired.fastq"
RU2=$ABBREV"_R2_unpaired.fastq"

THREADS=1  #Line 9 sets this up to run 20 processors.  If you want fewer, make sure to change that line as well as this one.

refgenome=mBra_mito_full.fa  	# your reference genome for the assembly
REF_HOME=$BASEDIR/reference	#the location of your reference genome

RAW_READS_HOME=/lustre/scratch/daray/uce.partition1/fastq   #the location of your raw data
mkdir $BASEDIR/data_raw
UNZIPPED_RAW_HOME=$BASEDIR/data_raw
mkdir $BASEDIR/data_processed
PROCESSED_READS_HOME=$BASEDIR/data_processed	#the location of your processed data
mkdir $PROCESSED_READS_HOME/quality_info
QUALITY_INFO=$PROCESSED_READS_HOME/quality_info	#Where the quality stats will be saved
mkdir $BASEDIR/support_files
SUPPORT_FILES=$BASEDIR/support_files	#where the support files like the adapter sequences will be located.

######
#set up alias' for major programs
######
BWA_HOME=/lustre/work/apps/bwa-0.6.2
SAMTOOLS_HOME=/lustre/work/apps/samtools-1.2
SAMTOOLS1_8_HOME=/lustre/work/apps/samtools-0.1.18
PICARD_HOME=/lustre/work/apps/picard-tools-1.91
BCFTOOLS_HOME=/lustre/work/apps/samtools-0.1.18/bcftools
RAY_SOFTWARE=/lustre/work/daray/software
TRIM_HOME=/lustre/work/apps/Trimmomatic-0.27
FASTX_HOME=/lustre/work/apps/fastx_toolkit-0.0.14/bin
VCFTOOLS_HOME=/lustre/work/daray/software/vcftools_0.1.12b/bin
MIRA_HOME=/lustre/work/apps/mira


#======================
#MIRA4 assembly 
#Unzip the raw reads into the processed_reads directory
gunzip -c $RAW_READS_HOME/$ABBREV"_L001_R1_001.fastq.gz" >$UNZIPPED_RAW_HOME/$ABBREV"L001_R1_001.fastq"
gunzip -c $RAW_READS_HOME/$ABBREV"_L001_R2_001.fastq.gz" >$UNZIPPED_RAW_HOME/$ABBREV"L001_R2_001.fastq"

$MIRA_HOME/bin/mira $SUPPORT_FILES/07LEP_manifest.conf

#Bait and iteratively map to the reference genome
perl $RAY_SOFTWARE/MITObim_1.8.pl \
	-start 1 \
	-end 10 \
	-sample 07REP \
	-ref mBra \
	-readpool reads.fastq \
	-maf 07LEP_mito_assembly/07LEP_mito_d_results/07LEP_mito_out.maf \
	&> $ABBREV"_log"


-bash-4.1$ /PATH/TO/MITObim.pl -start 1 -end 10 -sample testpool -ref Salpinus_mt_genome -readpool reads.fastq -maf initial-mapping-testpool-to-Salpinus-mt_assembly/initial-mapping-testpool-to-Salpinus-mt_d_results/initial-mapping-testpool-to-Salpinus-mt_out.maf &> log






