#!/bin/bash
#$ -V
#$ -cwd
#$ -S /bin/bash
#$ -N QC
#$ -o $JOB_NAME.o$JOB_ID
#$ -e $JOB_NAME.e$JOB_ID
#$ -q Chewie
#$ -pe fill 20 
#$ -P communitycluster

#This file will process raw Illumina data using Trimmomatic.  This will be followed by mapping to a reference genome to create a new genome assembly.

BASEDIR=/lustre/scratch/daray/Ray_low_cov_work/transcript1
WORKDIR=$BASEDIR/processed_reads

cd $WORKDIR

THREADS=19  #Line 9 sets this up to run 20 processors.  If you want fewer, make sure to change that line as well as this one.

RAW_READS_HOME=$BASEDIR/reads   #the location of your raw data
PROCESSED_READS_HOME=$BASEDIR/processed_reads	#the location of your processed data
QUALITY_INFO=$PROCESSED_READS_HOME/QC_files	#Where the quality stats will be saved
SUPPORT_FILES=$BASEDIR/support_files	#where the support files like the adapter sequences will be located.

######
#set up alias' for major programs
######
BWA_HOME=/lustre/work/apps/bwa-0.6.2
SAMTOOLS_HOME=/lustre/work/apps/samtools-1.2
SAMTOOLS1_8_HOME=/lustre/work/apps/samtools-0.1.18
PICARD_HOME=/lustre/work/apps/picard-tools-1.91
BCFTOOLS_HOME=/lustre/work/apps/samtools-0.1.18/bcftools
RAY_SOFTWARE=/lustre/work/daray/software
TRIM_HOME=/lustre/work/apps/Trimmomatic-0.27
FASTX_HOME=/lustre/work/apps/fastx_toolkit-0.0.14/bin
VCFTOOLS_HOME=/lustre/work/daray/software/vcftools_0.1.12b/bin
BEDTOOLS_HOME=/lustre/work/apps/bedtools-2.17.0/bin

#for RAW_READS_GZ in $RAW_READS_HOME/*.fastq.gz
#do gunzip $RAW_READS_GZ
#done

for RAW_READ_FILE in $RAW_READS_HOME/*_R1.fastq
do ABBREV=$(basename $RAW_READ_FILE _R1.fastq) 

#These files will be generated by trimmomatic
RP1=$ABBREV"_R1_paired.fastq"
RP2=$ABBREV"_R2_paired.fastq"
RU1=$ABBREV"_R1_unpaired.fastq"
RU2=$ABBREV"_R2_unpaired.fastq"


###########
#Prepare the reads and get quality stats
###########
#[1] Use trimmomatic to generate paired and unpaired reads files
java -jar $TRIM_HOME/trimmomatic-0.27.jar \
	PE \
	-threads $THREADS \
	-phred33 \
	$RAW_READS_HOME/$ABBREV"_R1.fastq" \
	$RAW_READS_HOME/$ABBREV"_R2.fastq" \
	$PROCESSED_READS_HOME/$RP1 \
	$PROCESSED_READS_HOME/$RU1 \
	$PROCESSED_READS_HOME/$RP2 \
	$PROCESSED_READS_HOME/$RU2 \
	ILLUMINACLIP:$SUPPORT_FILES/TruSeq4-PE.fa:2:30:10 \
	LEADING:20 \
	TRAILING:20 \
	SLIDINGWINDOW:4:20 \
	MINLEN:33
	
cat $PROCESSED_READS_HOME/$RU1 \
	$PROCESSED_READS_HOME/$RU2 \
	>$PROCESSED_READS_HOME/$ABBREV"_RX_unpaired.fastq"

done

#[2] Generate quality stats 
for FASTQ in $PROCESSED_READS_HOME/$ABBREV*.fastq
	do 	PROCESSED_FILE=$(basename $FASTQ .fastq)

$FASTX_HOME/fastx_quality_stats 					\
	-Q33		 						\
	-o $QUALITY_INFO/$PROCESSED_FILE".stats" 	\
	-i $PROCESSED_READS_HOME/$PROCESSED_FILE".fastq"

$FASTX_HOME/fastx_nucleotide_distribution_graph.sh 			\
	-i $QUALITY_INFO/$PROCESSED_FILE".stats"		\
	-o $QUALITY_INFO/$PROCESSED_FILE"_NUC.png"	\
	-t $QUALITY_INFO/$PROCESSED_FILE"_clipped"		

$FASTX_HOME/fastq_quality_boxplot_graph.sh 				\
	-i $QUALITY_INFO/$PROCESSED_FILE".stats" 	\
	-o $QUALITY_INFO/$PROCESSED_FILE"_BOX.png"	\
	-t $QUALITY_INFO/$PROCESSED_FILE"_clipped"

done

echo "_qc_finished" |  mailx -s "_qc_finished" david.4.ray@gmail.com

sleep 5

